src/3d/qgsoffscreen3dengine.cpp

```diff
@@ -0,0 +1,230 @@
+
+}
+
+QgsOffscreen3DEngine::~QgsOffscreen3DEngine()
+{
+  delete mAspectEngine;
+}
+
+void QgsOffscreen3DEngine::setSize( const QSize &s )
+{
+  mSize = s;
+
+  mTexture->setSize( mSize.width(), mSize.height() );
+  mDepthTexture->setSize( mSize.width(), mSize.height() );
+  mSurfaceSelector->setExternalRenderTargetSize( mSize );
+
+  mCamera->setAspectRatio( float( mSize.width() ) / float( mSize.height() ) );
+}
+
+void QgsOffscreen3DEngine::setClearColor( const QColor &color )
+{
+  mClearBuffers->setClearColor( color );
+}
+
+void QgsOffscreen3DEngine::setFrustumCullingEnabled( bool enabled )
+{
+  // TODO
+  Q_UNUSED( enabled );
+}
+
+void QgsOffscreen3DEngine::createRenderTarget()
+{
+  mTextureTarget = new Qt3DRender::QRenderTarget;
+
+  // The lifetime of the objects created here is managed
+  // automatically, as they become children of this object.
+
+  // Create a render target output for rendering colour.
+  mTextureOutput = new Qt3DRender::QRenderTargetOutput( mTextureTarget );
+  mTextureOutput->setAttachmentPoint( Qt3DRender::QRenderTargetOutput::Color0 );
+
+  // Create a texture to render into.
+  mTexture = new Qt3DRender::QTexture2D( mTextureOutput );
+  mTexture->setSize( mSize.width(), mSize.height() );
+  mTexture->setFormat( Qt3DRender::QAbstractTexture::RGB8_UNorm );
+  mTexture->setMinificationFilter( Qt3DRender::QAbstractTexture::Linear );
+  mTexture->setMagnificationFilter( Qt3DRender::QAbstractTexture::Linear );
+
+  // Hook the texture up to our output, and the output up to this object.
+  mTextureOutput->setTexture( mTexture );
+  mTextureTarget->addOutput( mTextureOutput );
+
+  mDepthTextureOutput = new Qt3DRender::QRenderTargetOutput( mTextureTarget );
+  mDepthTextureOutput->setAttachmentPoint( Qt3DRender::QRenderTargetOutput::Depth );
+  mDepthTexture = new Qt3DRender::QTexture2D( mDepthTextureOutput );
+  mDepthTexture->setSize( mSize.width(), mSize.height() );
+  mDepthTexture->setFormat( Qt3DRender::QAbstractTexture::DepthFormat );
+  mDepthTexture->setMinificationFilter( Qt3DRender::QAbstractTexture::Linear );
+  mDepthTexture->setMagnificationFilter( Qt3DRender::QAbstractTexture::Linear );
+  mDepthTexture->setComparisonFunction( Qt3DRender::QAbstractTexture::CompareLessEqual );
+  mDepthTexture->setComparisonMode( Qt3DRender::QAbstractTexture::CompareRefToTexture );
+  // Hook up the depth texture
+  mDepthTextureOutput->setTexture( mDepthTexture );
+  mTextureTarget->addOutput( mDepthTextureOutput );
+}
+
+void QgsOffscreen3DEngine::createFrameGraph()
+{
+  // Firstly, create the offscreen surface. This will take the place
+  // of a QWindow, allowing us to render our scene without one.
+  mOffscreenSurface = new QOffscreenSurface();
+  mOffscreenSurface->setFormat( QSurfaceFormat::defaultFormat() );
+  mOffscreenSurface->create();
+
+  // Hook it up to the frame graph.
+  mSurfaceSelector = new Qt3DRender::QRenderSurfaceSelector( mRenderSettings );
```

Comment:
Leaked?

Reply:
Checked and this should get parented to mRenderSettings (added a comment about it)